src/main/kotlin/com/avail/optimizer/values/L2SemanticPrimitiveInvocation.kt

```diff
@@ -0,0 +1,145 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice, this
+ *   list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of the copyright holder nor the names of the contributors
+ *   may be used to endorse or promote products derived from this software
+ *   without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.avail.optimizer.values
+
+import com.avail.descriptor.representation.AvailObject
+import com.avail.interpreter.Primitive
+import com.avail.utility.Casts
+
+/**
+ * An [L2SemanticValue] which represents the result produced by a [Primitive]
+ * when supplied a list of argument [L2SemanticValue]s.  The primitive must be
+ * stable (same result), pure (no side-effects), and successful for the supplied
+ * arguments.
+ *
+ * @author Mark van Gulik &lt;mark@availlang.org&gt;
+ * @author Todd L Smith &lt;todd@availlang.org&gt;
+ *
+ * @constructor
+ * Create a new `L2SemanticPrimitiveInvocation` semantic value.
+ *
+ * @property primitive
+ *   The [Primitive] whose invocation is being represented.
+ * @property argumentSemanticValues
+ *   The [List] of [L2SemanticValue]s that represent the arguments to the
+ *   invocation of the primitive.
+ * @param primitive
+ *   The primitive whose invocation is being represented.
+ * @param argumentSemanticValues
+ *   The semantic values supplied as arguments.
+ */
+class L2SemanticPrimitiveInvocation internal constructor(
+	@JvmField val primitive: Primitive,
+	@JvmField val argumentSemanticValues: List<L2SemanticValue>)
+		: L2SemanticValue(computeHash(primitive, argumentSemanticValues))
+{
+	init
+	{
+		assert(primitive.hasFlag(Primitive.Flag.CanFold))
+	}
+
+	override fun equals(obj: Any?): Boolean
+	{
+		if (obj !is L2SemanticPrimitiveInvocation)
+		{
+			return false
+		}
+		val invocation =
+			Casts.cast<Any, L2SemanticPrimitiveInvocation>(obj)
```

Comment:
This is a little shorter as:
```
val invocation = obj as L2SemanticPrimitiveInvocation
```

Reply:
actually, this entire variable isn't needed because of the conditional check on line 71. Obj doesn't need to be cast at this point.